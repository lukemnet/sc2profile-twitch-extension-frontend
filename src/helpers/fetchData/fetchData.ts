import { FetchParams } from 'types';

const fetchData = async ({
  url,
  method,
  headers,
}: FetchParams) => {
  try {
    // I'm not exactly happy about this hack
    // but I'm desperate to prevent unnecessary fetches
    // before I figure out what's with unintended rerenders
    if (!(headers as any).channelId) {
      return Promise.resolve({});
    }

    const response = await fetch(url, {
      method: method || 'GET',
      headers,
    });
    return await response.json();
  } catch {
    return {
      error: true,
      status: 500,
    };
  }
};

export default fetchData;
