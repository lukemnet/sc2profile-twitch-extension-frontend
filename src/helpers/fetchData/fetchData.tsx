import { FetchParams } from 'types';

const fetchData = async ({
  url,
  method,
  headers,
}: FetchParams) => {
  try {
    const response = await fetch(url, {
      method: method || 'GET',
      headers,
    });
    return await response.json();
  } catch {
    return {
      error: true,
      status: 500,
    };
  }
};

export default fetchData;
