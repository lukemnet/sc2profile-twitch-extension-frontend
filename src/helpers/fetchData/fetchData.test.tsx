import fetchData from './fetchData';

beforeEach(() => {
  global.fetch.resetMocks();
});

it('fetches data correctly when no method is provided', async () => {
  global.fetch.mockResponseOnce(JSON.stringify({
    status: 200,
    message: 'Mocked data for fetchData "GET" request',
  }));

  const response = await fetchData({
    url: 'http://example.com',
  });

  expect(response).toMatchSnapshot();
});
it('fetches data correctly for "GET" request', async () => {
  global.fetch.mockResponseOnce(JSON.stringify({
    status: 200,
    message: 'Mocked data for fetchData "GET" request',
  }));

  const response = await fetchData({
    url: 'http://example.com',
    method: 'GET',
  });

  expect(response).toMatchSnapshot();
});

it('fetches data correctly for "POST" request', async () => {
  global.fetch.mockResponseOnce(JSON.stringify({
    status: 200,
    message: 'Mocked data for fetchData "POST" request',
  }));

  const response = await fetchData({
    url: 'http://example.com',
    method: 'POST',
  });

  expect(response).toMatchSnapshot();
});

it('returns error object when request fails', async () => {
  global.fetch.mockResponseOnce(Error('mocked_fetch_error'));

  const response = await fetchData({
    url: 'http://example.com',
    method: 'POST',
  });

  expect(response).toMatchSnapshot();
});