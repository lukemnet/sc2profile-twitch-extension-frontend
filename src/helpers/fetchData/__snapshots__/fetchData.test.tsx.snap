// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches data correctly for "GET" request 1`] = `
Object {
  "message": "Mocked data for fetchData \\"GET\\" request",
  "status": 200,
}
`;

exports[`fetches data correctly for "POST" request 1`] = `
Object {
  "message": "Mocked data for fetchData \\"POST\\" request",
  "status": 200,
}
`;

exports[`fetches data correctly when no method is provided 1`] = `
Object {
  "message": "Mocked data for fetchData \\"GET\\" request",
  "status": 200,
}
`;

exports[`returns error object when request fails 1`] = `
Object {
  "error": true,
  "status": 500,
}
`;
